opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,249 @@
+            return
+        else:
+            data = {**data, **value}
+
+        if self.is_room_encrypted and self.should_encrypt:
+            room_event = await self.opsdroid.send(
+                GenericMatrixRoomEvent(
+                    target=self.room_id, content=value, event_type=self._event_type
+                )
+            )
+            data[key] = {"encrypted_val": room_event.event_id}
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id} with {data}")
+
+        await self.opsdroid.send(
+            MatrixStateEvent(
+                self._event_type,
+                content=data,
+                target=self.room_id,
+                connector=self.connector,
+                state_key=state_key,
+            )
+        )
+
+        return True
+
+    async def get(self, key, get_full=False):
+        """Get a value from the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        _LOGGER.debug(
+            f"Getting {key} from matrix room {self.room_id} with state_key={state_key}."
+        )
+
+        ori_data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(ori_data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error getting {key} from matrix room {self.room_id}: {ori_data.message}({ori_data.status_code})"
+            )
+            return
+        elif ori_data.transport_response.status == 404:
+            _LOGGER.error(
+                f"Error getting {key} from matrix room {self.room_id}: Event not found"
+            )
+            return
+
+        data = ori_data.content
+
+        _LOGGER.debug(f"Got {data} from state request.")
+
+        if self._single_state_key:
+            if key in data:
+                data = {key: data[key]}
+            elif not get_full:
+                _LOGGER.debug(f"{key} doesn't exist in database")
+                return
+            else:
+                return data
+
+        for k, v in data.items():
+            if isinstance(v, dict) and "encrypted_val" in v:
+                resp = await self.connector.connection.room_get_event(
+                    room_id=self.room_id, event_id=v["encrypted_val"],
+                )
+                if isinstance(resp, RoomGetEventError):
+                    _LOGGER.error(
+                        f"Error decrypting {v['encrypted_val']} while getting "
+                        f"{key}: {resp.message}({resp.status_code})"
+                    )
+                    continue
+                data[k] = resp.event.source["content"][k]
+
+        if get_full:
+            return {**ori_data.content, **data}
```

Comment:
Doing

```
data = ori_data.content
```
above did a copy by reference, which means that assigning to `data[k]` has modified `ori_data.content` so I am not sure if this line is doing anything? (i.e. return data would be the same).

Reply:
It's to account for the reassignment in [line 162](https://github.com/opsdroid/opsdroid/pull/1587/files#diff-c97d0db15ae0cc89347831db7cc93d05R162) so that when ssk is true we send the entire dict back to put